In [ ]:
import numpy as np
import pandas as pd

In [ ]:
!mkdir -p ~/.kaggle
!cp kaggle.json ~/.kaggle/

In [ ]:
!kaggle datasets download -d tmdb/tmdb-movie-metadata

 90% 8.00M/8.89M [00:01<00:00, 12.6MB/s]
100% 8.89M/8.89M [00:01<00:00, 7.62MB/s]


In [ ]:
import zipfile
zip_ref = zipfile.ZipFile('/content/tmdb-movie-metadata.zip', 'r')
zip_ref.extractall('/content')
zip_ref.close()

In [ ]:
data1=pd.read_csv("/content/tmdb_5000_movies.csv")
data1.head(1)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   

  original_title                                           overview  \
0         Avatar  In the 22nd century, a paraplegic Marine is di...   

   popularity                               production_companies  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   

                       tagline   title  vote_average  vote_count  
0  Enter the World of Pandora.  Avatar           7.2       11800

In [ ]:
data2=pd.read_csv("/content/tmdb_5000_credits.csv")
data2.head(1)

movie_id   title                                               cast  \
0     19995  Avatar  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

In [ ]:
data1=data1.merge(data2,on="title")
#data.shape
data1.head(1)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   

  original_title                                           overview  \
0         Avatar  In the 22nd century, a paraplegic Marine is di...   

   popularity                               production_companies  ... runtime  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...  ...   162.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   

                       tagline   title vote_average vote_count movie_id  \
0  Enter the World of Pandora.  Avatar          7.2      11800    19995   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  

[1 rows x 23 columns]

### Features Extraction
- movie_id
- tittle
- genres
- keywords
- overview
- cast
- crew

In [ ]:
data1=data1[["movie_id","genres","title","keywords","overview","cast","crew"]]
#data.head(2)
data1.shape

(4809, 7)

In [ ]:
data1.isnull().sum()

movie_id    0
genres      0
title       0
keywords    0
overview    3
cast        0
crew        0
dtype: int64

In [ ]:
data1.dropna(inplace=True)

In [ ]:
data1.duplicated().sum()

0

In [ ]:
data1.iloc[0].genres

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

#### why use ast: {
- import ast
- ast.literal_eval('[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]')
}

In [ ]:
import ast
def convert(obj):
  new_list=[]
  for i in ast.literal_eval(obj):
    new_list.append(i["name"])
  return new_list

In [ ]:
#convert('[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]')

data1["genres"]=data1["genres"].apply(convert)

In [ ]:
data1["genres"][0:5]

0    [Action, Adventure, Fantasy, Science Fiction]
1                     [Adventure, Fantasy, Action]
2                       [Action, Adventure, Crime]
3                 [Action, Crime, Drama, Thriller]
4             [Action, Adventure, Science Fiction]
Name: genres, dtype: object

In [ ]:

data1["keywords"][0]

'[{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "space war"}, {"id": 3388, "name": "space colony"}, {"id": 3679, "name": "society"}, {"id": 3801, "name": "space travel"}, {"id": 9685, "name": "futuristic"}, {"id": 9840, "name": "romance"}, {"id": 9882, "name": "space"}, {"id": 9951, "name": "alien"}, {"id": 10148, "name": "tribe"}, {"id": 10158, "name": "alien planet"}, {"id": 10987, "name": "cgi"}, {"id": 11399, "name": "marine"}, {"id": 13065, "name": "soldier"}, {"id": 14643, "name": "battle"}, {"id": 14720, "name": "love affair"}, {"id": 165431, "name": "anti war"}, {"id": 193554, "name": "power relations"}, {"id": 206690, "name": "mind and soul"}, {"id": 209714, "name": "3d"}]'

In [ ]:
data1["keywords"]=data1["keywords"].apply(convert)

In [ ]:
data1["keywords"][0]

['culture clash',
 'future',
 'space war',
 'space colony',
 'society',
 'space travel',
 'futuristic',
 'romance',
 'space',
 'alien',
 'tribe',
 'alien planet',
 'cgi',
 'marine',
 'soldier',
 'battle',
 'love affair',
 'anti war',
 'power relations',
 'mind and soul',
 '3d']

In [ ]:
data1.head(2)

movie_id                                         genres  \
0     19995  [Action, Adventure, Fantasy, Science Fiction]   
1       285                   [Adventure, Fantasy, Action]   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...

In [ ]:
import ast
def convert3(obj):
  new_list=[]
  count=1
  for i in ast.literal_eval(obj):
    if count!=4:
      new_list.append(i["name"])
      count+=1
    else:
      break
  return new_list

In [ ]:
data1["cast"]=data1["cast"].apply(convert3)

In [ ]:
data1.head(2)

movie_id                                         genres  \
0     19995  [Action, Adventure, Fantasy, Science Fiction]   
1       285                   [Adventure, Fantasy, Action]   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   

                                               cast  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]   
1     [Johnny Depp, Orlando Bloom, Keira Knightley]   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...

In [ ]:
data1["crew"][0]

'[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cameron"},

In [ ]:
import ast
def find_director_name(obj):
  new_list=[]
  for i in ast.literal_eval(obj):
    if i["job"]=="Director":
      new_list.append(i["name"])
      break
  return new_list

In [ ]:
data1["crew"]=data1["crew"].apply(find_director_name)

In [ ]:
data1["crew"]

0           [James Cameron]
1          [Gore Verbinski]
2              [Sam Mendes]
3       [Christopher Nolan]
4          [Andrew Stanton]
               ...         
4804     [Robert Rodriguez]
4805         [Edward Burns]
4806          [Scott Smith]
4807          [Daniel Hsia]
4808     [Brian Herzlinger]
Name: crew, Length: 4806, dtype: object

In [ ]:
data1.head(1)

movie_id                                         genres   title  \
0     19995  [Action, Adventure, Fantasy, Science Fiction]  Avatar   

                                            keywords  \
0  [culture clash, future, space war, space colon...   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   

                                               cast             crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]  [James Cameron]

In [ ]:
data1["overview"]=data1["overview"].apply(lambda x:x.split())

In [ ]:
data1.head(2)

movie_id                                         genres  \
0     19995  [Action, Adventure, Fantasy, Science Fiction]   
1       285                   [Adventure, Fantasy, Action]   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   

                                               cast              crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]   [James Cameron]  
1     [Johnny Depp, Orlando Bloom, Keira Knightley]  [Gore Verbinski]

#### remove space between these

In [ ]:
data1['keywords']=data1['keywords'].apply(lambda x:[i.replace(" ","") for i in x])
data1['genres']=data1['genres'].apply(lambda x:[i.replace(" ","") for i in x])
data1['cast']=data1['cast'].apply(lambda x:[i.replace(" ","") for i in x])
data1['crew']=data1['crew'].apply(lambda x:[i.replace(" ","") for i in x])


In [ ]:
data1.head(1)

movie_id                                        genres   title  \
0     19995  [Action, Adventure, Fantasy, ScienceFiction]  Avatar   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   

                                            cast            crew  
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]  [JamesCameron]

In [ ]:
data1["tags"]=data1["genres"]+data1["keywords"]+data1["overview"]+data1["cast"]+data1["crew"]

In [ ]:
new_data=data1[["movie_id","title","tags"]]

In [ ]:
new_data.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  [Action, Adventure, Fantasy, ScienceFiction, c...  
1  [Adventure, Fantasy, Action, ocean, drugabuse,...  
2  [Action, Adventure, Crime, spy, basedonnovel, ...  
3  [Action, Crime, Drama, Thriller, dccomics, cri...  
4  [Action, Adventure, ScienceFiction, basedonnov...

#### convert tags into string

In [ ]:
new_data["tags"]=new_data["tags"].apply(lambda x:" ".join(x))

<ipython-input-35-64b032f56e7c>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_data["tags"]=new_data["tags"].apply(lambda x:" ".join(x))


In [ ]:
new_data.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  Action Adventure Fantasy ScienceFiction cultur...  
1  Adventure Fantasy Action ocean drugabuse exoti...  
2  Action Adventure Crime spy basedonnovel secret...  
3  Action Crime Drama Thriller dccomics crimefigh...  
4  Action Adventure ScienceFiction basedonnovel m...

#### convert tags into lower case

In [ ]:
new_data["tags"]=new_data["tags"].apply(lambda x:x.lower())

<ipython-input-37-a8964160f85f>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_data["tags"]=new_data["tags"].apply(lambda x:x.lower())


In [ ]:
new_data["tags"][0]

'action adventure fantasy sciencefiction cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d in the 22nd century, a paraplegic marine is dispatched to the moon pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. samworthington zoesaldana sigourneyweaver jamescameron'

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
cv=CountVectorizer(max_features=5000,stop_words="english")

In [ ]:
vectors=cv.fit_transform(new_data["tags"]).toarray()

In [ ]:
vectors

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [ ]:
vectors.shape

(4806, 5000)

In [ ]:
len(cv.get_feature_names_out())

5000

#### apply stemming

In [ ]:
import nltk

In [ ]:
from nltk.stem.porter import PorterStemmer
ps=PorterStemmer()

In [ ]:
def stem(text):
  y=[]
  for i in text.split():
    y.append(ps.stem(i))
  return " ".join(y)

In [ ]:
stem('action adventure fantasy sciencefiction cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d in the 22nd century, a paraplegic marine is dispatched to the moon pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. samworthington zoesaldana sigourneyweaver jamescameron')

'action adventur fantasi sciencefict cultureclash futur spacewar spacecoloni societi spacetravel futurist romanc space alien tribe alienplanet cgi marin soldier battl loveaffair antiwar powerrel mindandsoul 3d in the 22nd century, a parapleg marin is dispatch to the moon pandora on a uniqu mission, but becom torn between follow order and protect an alien civilization. samworthington zoesaldana sigourneyweav jamescameron'

In [ ]:
new_data["tags"]=new_data["tags"].apply(stem)

<ipython-input-48-bc497b5cc05b>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_data["tags"]=new_data["tags"].apply(stem)


In [ ]:
new_data.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  action adventur fantasi sciencefict culturecla...  
1  adventur fantasi action ocean drugabus exotici...  
2  action adventur crime spi basedonnovel secreta...  
3  action crime drama thriller dccomic crimefight...  
4  action adventur sciencefict basedonnovel mar m...

#### find the distance between the vectors   
- cosine distance

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity


In [ ]:
similarity=cosine_similarity(vectors)

In [ ]:
similarity

array([[1.        , 0.08964215, 0.05976143, ..., 0.02519763, 0.02817181,
        0.        ],
       [0.08964215, 1.        , 0.0625    , ..., 0.02635231, 0.        ,
        0.        ],
       [0.05976143, 0.0625    , 1.        , ..., 0.02635231, 0.        ,
        0.        ],
       ...,
       [0.02519763, 0.02635231, 0.02635231, ..., 1.        , 0.0745356 ,
        0.04836508],
       [0.02817181, 0.        , 0.        , ..., 0.0745356 , 1.        ,
        0.05407381],
       [0.        , 0.        , 0.        , ..., 0.04836508, 0.05407381,
        1.        ]])

In [ ]:
similarity.shape

(4806, 4806)

#### first movie similarity with other 4806 movie  the ---> **array([1.** <---      denotes the similarity of 1st with 1st

In [ ]:
similarity[0]

array([1.        , 0.08964215, 0.05976143, ..., 0.02519763, 0.02817181,
       0.        ])

In [ ]:
similarity[0].shape

(4806,)

In [ ]:
movie_index=new_data[new_data["title"]=="Spectre"].index[0]
movie_index

2

In [ ]:
sorted(list(enumerate(similarity[0])),reverse=True, key=lambda x:x[1])[1:6]

[(539, 0.26089696604360174),
 (1194, 0.2581988897471611),
 (507, 0.25302403842552984),
 (260, 0.25110592822973776),
 (1216, 0.24944382578492943)]

In [ ]:
def recommend(movie):
  movie_index=new_data[new_data["title"]==movie].index[0]
  distance=similarity[movie_index]
  movie_list=sorted( list (enumerate(distance)),reverse=True, key=lambda x: x[1])[1:6]

  for i in movie_list:
    print(new_data.iloc[i[0]].title)

In [ ]:
recommend('Gandhi')

Gandhi, My Father
The Wind That Shakes the Barley
A Passage to India
Guiana 1838
Ramanujan


In [ ]:
!pip install pickle-mixin

  Preparing metadata (setup.py) ... done
  Created wheel for pickle-mixin: filename=pickle_mixin-1.0.2-py3-none-any.whl size=5990 sha256=3293bdcec5ed1bf419b8c7bd891e5ab3bc9ad4194cb602d3ea926a2e4bd6f869
  Stored in directory: /root/.cache/pip/wheels/3e/c6/e9/d1b0a34e1efc6c3ec9c086623972c6de6317faddb2af0a619c
Successfully built pickle-mixin


In [ ]:
import pickle

In [ ]:
pickle.dump(new_data,open("movies.pkl",'wb'))

In [ ]:
pickle.dump(new_data.to_dict(),open("movies_dict.pkl",'wb'))

In [ ]:
pickle.dump(similarity,open("similarity.pkl",'wb'))